Call NewSpanExporter without checking OTLP_TRACES_EXPORTER environment variable

ManiaKazemzadeh opened the pull request:
Currently if `OTEL_TRACES_EXPORTER` is not defined, then the trace exporter is always set to HTTP, regardless of the value of `OTEL_EXPORTER_OTLP_PROTOCOL`.  This is because there is a check to see if `OTEL_TRACES_EXPORTER` is set before calling `autoexport.NewSpanExporter`.

This change removes that check and adds a test to ensure that `autoexport.NewSpanExporter` still works without `OTEL_TRACES_EXPORTER` defined. I have left the fallback code in place [here](https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/main/instrumentation.go#L216-L221) so that if `OTEL_EXPORTER_OTLP_PROTOCOL` has an invalid value (or NewSpanExporter returns nil for any other reason) the http exporter will be selected (which is the current behaviour).

MrAlias commented on instrumentation_test.go:
```diff
@@ -95,6 +96,16 @@ func TestWithEnv(t *testing.T) {
+		c, err := newInstConfig(context.Background(), []InstrumentationOption{WithEnv()})
+
+		require.NoError(t, err)
+		assert.NotNil(t, c.traceExp)
```
This should validate the instance type to be and exporter from `otlptracegrpc`. Given the structure of those exporters, calling [`MarshalLog`](https://pkg.go.dev/go.opentelemetry.io/otel/exporters/otlp/otlptrace#Exporter.MarshalLog) is likely your best option.

```suggestion
		require.NotNil(t, c.traceExp)
		require.IsType(t, &otlptrace.Exporter{}, c.traceExp)
		exp := c.traceExp.(*otlptrace.Exporter)

		var buf bytes.Buffer
		logger := stdr.New(log.New(&buf, "", log.LstdFlags))
		logger.Info("", "exporter", exp)
		got, err := io.ReadAll(&buf)
		require.NoError(t, err)
		// "otlphttphttp" should be "otlptracegrpc". This is a bug upstream.
		assert.Contains(t, string(got), "otlphttphttp")
```

MrAlias replied:
Will be impacted by https://github.com/open-telemetry/opentelemetry-go/pull/6143. We can follow up with that update later though.

MrAlias commented:
Please be sure to include a changelog entry describing the fix being applied.

ManiaKazemzadeh commented on instrumentation.go:
```diff
@@ -408,14 +405,11 @@ func WithEnv() InstrumentationOption {
-			// a way to just pass the environment value currently. Just use
-			// NewSpanExporter which will re-read this value.
-
+		{
```
I'd added this to limit the scope of the `e error` variable but I'm happy to remove it :)

ManiaKazemzadeh commented on instrumentation_test.go:
```diff
@@ -95,6 +96,16 @@ func TestWithEnv(t *testing.T) {
+	// Test that autoexport.NewSpanExporter works when OTEL_TRACES_EXPORTER is
+	// not set
+	t.Run("With OTEL_TRACES_EXPORTER not set", func(t *testing.T) {
+		os.Unsetenv("OTEL_TRACES_EXPORTER")
```
Added :)

ManiaKazemzadeh commented:
Hey, thank you for the review! I've added your suggestions. I spent some time trying to understand and see if there was another way of testing the exporter type but your suggestion to use MarshalLog seems to be the best solution. I've also added a changelog item. Please let me know if there are any other changes you'd like me to make 👍
